### ISSUE TITLE
Segmentation fault seems to be caused by event_file_writer.py threading

### ISSUE THREAD

**Issue by Pet222** [tensorflow/tensorboard#6113]
Hi!

I am using tensorboard  2.10.1 with pytorch (SummaryWriter). After a while my code crashes and reports segmentation fault. The python faulthandler lists the threads started by the event file writer as a root cause and it seems they bootstrap (maybe due to the queue issue?). Unfortunately I have no more info in the prints of the faulthandler what could cause the segmentation faults. 

Fatal Python error: Segmentation fault

Thread 0x00007fd9317fa700 (most recent call first):
  File "/home/pet/miniconda3/envs/idp_py3.9_new/lib/python3.9/threading.py", line 316 in wait
  File "/home/pet/miniconda3/envs/idp_py3.9_new/lib/python3.9/queue.py", line 180 in get
  File "/home/pet/miniconda3/envs/idp_py3.9_new/lib/python3.9/site-packages/tensorboard/summary/writer/event_file_writer.py", line 227 in run
  File "/home/pet/miniconda3/envs/idp_py3.9_new/lib/python3.9/threading.py", line 973 in _bootstrap_inner
  File "/home/pet/miniconda3/envs/idp_py3.9_new/lib/python3.9/threading.py", line 930 in _bootstrap

This is at line 227 in the evnet_file_writer.py:
                if queue_wait_duration > 0:
                    data = self._queue.get(True, queue_wait_duration)
                else:

Since I have no more info from the logs I can not tell more about the root cause.

Some additional info: 
- a lot of files are written parallel but so far it was not an issue and since then the amount of eventfiles were reduced
- only threads of the event file writer are reported by the faulthandler
- the number of open file limits were increased to 5000 (ulimit -n 50000)
- disk is not full
- python is 3.9 (but the one optimized for Intel)
- flushing is triggered automatically in every 1s but I also manually trigger flushing to disk manually as well multiple times (not sure if this could cause a problem...)

**Comment by yatbear** [comment]
Hi @Pet222,

I found a similar issue: https://github.com/pytorch/pytorch/issues/30651#issuecomment-1352451188 - in this comment it says the import order somehow triggered the issue (not sure why yet), did you happen to import `torchvision` before `SummaryWriter`? If so could you try changing the import order?

If the above is not the case, can you please provide a bit more information about how the SummaryWriter was used, preferably the relevant code snippet? Thanks!

**Comment by Pet222** [comment]
Hi @yatbear !

Thanks for the hint. I do not use torchvision (neither import it). However I am not sure by now (after days of investigation) how tensorboard relevant my issue is ... Once this segmentation fault hit sometimes I can not really start my application either and it shows thread issue + pointing at pandas. Sometimes to tqdm.....In my program I have not implemented any threads but it seems there is some issue with threading and I assume pandas and tqdm uses threads too.....On the other hand I can not exclude that tb filewriter cases sg with the threads than the pandas etc is just a consecutive. Error....."PYDEVD_THREAD_DUMP_ON_WARN_EVALUATION_TIMEOUT": "1" maybe this will provide more info... 

Some code snipped:
I made my own class to preprocess log data. 

1. I initialize the SummaryWriter accordingly and I will refer to it as self.tb_writer

class custom_logger(object):

    """
    init the logger; init the TB writer
    """
    def __init__(self, logger_path: str):
        """
        logger_path: the path to the logger directory where the logs shall be saved
        """

        self.logger_path = logger_path
        self.log_dashboards = {}
        self.tb_writer = SummaryWriter(logger_path, flush_secs=1)


2. once the logs are preprocessed and re-ordered and my custom dashboards are ready the following is called:

self.tb_writer.add_scalars(dashboard + " (" + k + ")", _log_stat[k], gv.i_episode)

where
- dashboard[k] is a string which is the name of the dashboard, _log_stat[k] is the value I would like to write to the dashboard and the gv.i_episode is integer.

3. And this is the called in a for loop (this add scalar below is not the native TB add scalar but my custom one to preprocess the data - meaning it is collected stored and rearranged in a dictionary)

gv.custom_logger.add_scalar(dashboard="Agents_" + gv.MODE + "/Internal_Advantage",
                                        scalar_dict={
                                            "agent_" + str(self.name_or_id) + "_cum" : float(a_t)
                                        },
                                        step=_cnt_z,
                                        aggregate_mode=["raw", "mean","variance"])
gv.custom_logger.add_scalar(dashboard="Agents_" + gv.MODE + "/Discount",
                                        scalar_dict={
                                            "agent_" + str(self.name_or_id) + "_cum" : float(delta)
                                        },
                                        step=_cnt_z,
                                        aggregate_mode=["raw", "mean", "variance"])     


I also flush the data manually but after the for loop....but it can happen that I call an other flush quite close timewise to the first flush.....so maybe the flush is not complete and an other flush is called or parallelly during a flush a data is written into the filewriter.

Some extra info: 
- I use WSL2 + python 3.9 optimized for Intel CPUs.
- if I add some prints in the for loop which runs 100 times (to have some extra info approximately where the segmentation fault occurs since the debugger is not really helpful or I am not really an expert of threading yet) I can not reproduce the issue.
-  I've read in python doc that there are some env flags to try to get more info about threading issues so I am running the code with: "PYDEVD_THREAD_DUMP_ON_WARN_EVALUATION_TIMEOUT": "1". It has been running almost for a day now without any issue. 

In case you have some idea how/what to check let me know!

**Comment by yatbear** [comment]
Hi @Pet222,

Thanks so much for the detailed information! 

One quick question - have you tried commenting out the Pandas import and data processing temporarily, and write some fake data in your look to test if there's still segmentation fault? For example:
```
for k in range(10000):  # Or your own loop if it doesn't depend on Pandas
  self.tb_writer.add_scalars("dashboard(" + k + ")", _log_stat[k], gv.i_episode)
  # Or use fake data if `_log_stat` is a Pandas dataframe
  # self.tb_writer.add_scalars("dashboard(" + k + ")", k**0.5, step=k)
```
Pandas is not thread safe (https://github.com/pandas-dev/pandas/issues/2728 - there is not recent updates on this issue), if disabling the Pandas code got rid of the segfault, maybe it's worth considering replacing the Pandas processing with NumPy equivalent?

**Comment by Pet222** [comment]
Hi @yatbear,

Yes I noticed that, too :( Thanks for the hint. 

When pandas failed at the start I made a threading lock around the piece of code which according to python caused a seg fault (with lock.acquire() and lock.release()).  After this lock mechanism was introduced python started....but again (with this modification) after a while I had the segmentation fault...

Actually I do not really use pandas during the execution. What I do is that at the very beginning of the code I read data from .csv and preprocess the whole dataset with pandas.....and as you wrote everything is converted to numpy array or torch tensor and they are used during the rest of the data collection and training....

...my problem is that I wish the issue could be so well and stable reproducible so that I can test what you have proposed. But it took nearly 24h now to get the segmentation fault again....unfortunately  PYDEVD_THREAD_DUMP_ON_WARN_EVALUATION_TIMEOUT provided no extra info...

**Comment by yatbear** [comment]
Hi @Pet222,

Have you tried loading the data with [numpy.genfromtxt](https://numpy.org/doc/stable/reference/generated/numpy.genfromtxt.html) instead of Pandas? And is the input data so big that it needs parallel processing there? Please let us know if there is a way to reproduce it more stably. Thanks!

**Comment by Pet222** [comment]
Hi! I think I've found the root cause which might not be SW related (at least not python, or tensorboard). It seems my motherboard overclocked the CPU which might have caused this issue (a bit strange since the CPU was 80 degree Celsius hot and all other params including frequencies seemed to be in spec). After turning off the CPU boosting in the motherboard's BIOS I was not able to reproduce the issue. When I turned it back on the threads were again messy after some hours....Therefore I close this point.
